RFC/src/RFC-0171_PeerToPeerMessagingProtocol.md

```diff
@@ -0,0 +1,662 @@
+- prune the routing table based on a filter criterion (e.g. last date seen)
+
+#### MessageDispatcher
+
+A MessageDispatcher associates MessageTypes to handlers. Each handler gets a MessageContext as a parameter.
+
+The MessageContext contains:
+- the requesting PeerConnection,
+- the MessageHeader
+- the deserialized message,
+- the OutboundMessageService.
+
+Basically, all the tools the handler needs to interact with the network.
+
+A MessageDispatcher is responsible for:
+- constructing the MessageContext
+- finding the message handler which is associated with the MessageType
+- passing the MessageContext to the handler
+- if the handler cannot be found, the message is ignored
+
+An example API may be:
+
+```rust,compile_fail
+let dispatcher = Dispatcher::new()
+    .middleware(logger)
+    .route(MessageType::BLOCKCHAIN::NEW_BLOCK, BlockHandlers::store_and_broadcast)
+    ...
+    .route(MessageType::NET::PING, send_pong)
+    .catch_all(catch_all_handler)
+    .finish();
+    
+inbound_msg_service.set_handler(dispatcher.handler);
+```
+
+#### InboundMessageService
+
+InboundMessageService is a service that receives messages over a non-blocking asynchronous socket and
+determines what to do with it. There are 3 options: handle, forward, discard.
+
+A pool of worker threads (with a configurable size) is started and each listen for messages on their $1:n$ `inproc` message
+socket. A `ZMQ_DEALER` socket is suggested for fair-queueing work amongst workers, who listen for work with a `ZMQ_REP`.
+The workers read off this socket and process the messages.
+
+An InboundMessageService:
+
+- MUST receive messages from all PeerConnections
+- MUST write the message to the worker socket
+
+A worker:
+
+- MUST deserialize the MessageHeader
+    - if unable to deserialize, MUST discard the message
+- MUST check the message signature
+    - SHOULD discard the message if the signature is invalid.
+    - MUST discard the message if the signature has been processed before
+- If the encryption flag is set:
+    - SHOULD attempt to decrypt the message
+        - if successful, process and handle the message
+        - otherwise, SHOULD forward the message using the `Closest` BroadcastStrategy 
+        - if the message is not encrypted, MUST discard it
+- If the destination [node ID] is set:
+    - if the destination match this node's ID: process and handle the message
+    - if the destination does not match: SHOULD forward the message using the `Closest` BroadcastStrategy 
+- If the destination is not set:
+    - if the MessageType is a kind of propagation message:
+        - MUST handle the message
+        - MUST forward the message using the `Closest` BroadcastStrategy,
+    - if the MessageType is a kind of encrypted message:
+        - MUST attempt to decrypt and handle the message
+        - if successful, SHOULD handle the message
+        - if unsuccessful, SHOULD forward the message using the `Closest` BroadcastStrategy,
```

Comment:
I'm thinking `Closest` could actually fail under certain topographies where you mostly have close neighbours in your routing table and they only have you. The message might never escape.

`Random` or `Flood` might be better options